ocaml/typing/typedecl.ml

```diff
@@ -1041,6 +1074,11 @@ let check_constraints env sdecl (_, decl) =
     sdecl.ptype_params decl.type_params;
   begin match decl.type_kind with
   | Type_abstract _ -> ()
+  (* We skip this check because with [or_null_reexport] the [type_kind]
+     does not match [sdecl.ptype_kind]. This is sound, since re-exporting
+     can't introduce new variables in the kind. *)
+  | Type_variant _ when
+    Builtin_attributes.has_or_null_reexport decl.type_attributes -> ()
```

Comment:
Annoyingly, this will foul the unused-attribute warning, because this function marks the attribute as used. You could get around this by matching for `Ptype_abstract` along with `Type_variant`. Or make a `check_or_null_reexport` that looks for the attribute but doesn't mark it as used? Or maybe just keep the check in `transl_declaration` that I said to remove, to prevent code from reaching here.

Reply:
Let's just keep the check.